File: sample-conference-videochat/sample-conference-videochat/Components/Authorization1/RootParentVC.m
//
//  RootParentVC.m
//  samplechat
//
//  Created by Vladimir Nybozhinsky on 1/29/20.
//  Copyright © 2020 Quickblox. All rights reserved.
//

#import "RootParentVC.h"
#import "SplashScreenVC.h"
#import "DialogsViewController.h"
#import "UIColor+Chat.h"
#import "ChatParentVC.h"

@interface DialogsNavigationController: UINavigationController
@end
@implementation DialogsNavigationController
@end

@interface RootParentVC ()
@property (nonatomic, strong) UIViewController *current;
@end

@implementation RootParentVC

- (instancetype)init
{
    self = [super init];
    if (self) {
        UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Authorization" bundle:nil];
        SplashScreenVC *splashScreenVC = [storyboard instantiateViewControllerWithIdentifier:@"SplashScreenVC"];
        self.current = splashScreenVC;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self addChildViewController:self.current];
    self.current.view.frame = self.view.bounds;
    [self.view addSubview:self.current.view];
    [self.current didMoveToParentViewController:self];
}

- (void)setDialogID:(NSString *)dialogID {
    _dialogID = dialogID;
    [self handlePush];
}

- (void)handlePush {
    if ([self.current isKindOfClass:[DialogsNavigationController class]] && self.dialogID) {
        DialogsNavigationController *dialogsNavigationController = (DialogsNavigationController *)self.current;
        
        self.dialogID = nil;
        
        NSMutableArray *newStack = [NSMutableArray array];
        NSArray *controllers = dialogsNavigationController.viewControllers;
        
        //change stack by replacing view controllers after CallViewController
        for (UIViewController *controller in controllers) {
            [newStack addObject:controller];
            
            if ([controller isKindOfClass:[DialogsViewController class]]) {
                [controller dismissViewControllerAnimated:NO completion:nil];
                NSArray *newControllers = [newStack copy];
                [dialogsNavigationController setViewControllers:newControllers];
                
                break;
            }
        }
        [dialogsNavigationController popToRootViewControllerAnimated:NO];
    }
}

- (void)showLoginScreen {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Authorization" bundle:nil];
    UINavigationController *new = [storyboard instantiateViewControllerWithIdentifier:@"AuthNavVC"];
    
    [self addChildViewController:new];
    new.view.frame = self.view.bounds;
    [self.view addSubview:new.view];
    [new didMoveToParentViewController:self];
    
    [self.current willMoveToParentViewController:nil];
    [self.current.view removeFromSuperview];
    [self.current removeFromParentViewController];
    self.current = new;
}

- (void)switchToDialogsScreen {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Dialogs" bundle:nil];
    DialogsViewController *dialogsScreen = [storyboard instantiateViewControllerWithIdentifier:@"DialogsViewController"];
    
    DialogsNavigationController *new = [[DialogsNavigationController alloc] initWithRootViewController:dialogsScreen];
    new.navigationBar.barTintColor = [UIColor mainColor];
    new.navigationBar.barStyle = UIBarStyleBlack;
    new.navigationBar.shadowImage = UIImage.new;
    [new.navigationBar setTranslucent:NO];
    new.navigationBar.tintColor = UIColor.whiteColor;
    new.navigationBar.titleTextAttributes = @{ NSForegroundColorAttributeName:UIColor.whiteColor};
    
    [self addChildViewController:new];
    new.view.frame = self.view.bounds;
    [self.view addSubview:new.view];
    [new didMoveToParentViewController:self];
    
    [self.current willMoveToParentViewController:nil];
    [self.current.view removeFromSuperview];
    [self.current removeFromParentViewController];
    self.current = new;
    
    [self handlePush];
}

- (void)animateFadeTransitionToNew:(UIViewController *)new
                        completion: (nullable void(^)(void))completion {
    
    [self addChildViewController:new];
    [new didMoveToParentViewController:self];
    [self.current willMoveToParentViewController:nil];
    [self.current.view removeFromSuperview];
    [self.current removeFromParentViewController];
    
    self.current = new;
    if (completion) {
        completion();
    }
}

- (void)animateDismissTransitionToNew:(UIViewController *)new
                           completion: (nullable void(^)(void))completion {
    
    CGRect initialFrame = CGRectMake(-self.view.bounds.size.width, 0.0, self.view.bounds.size.width, self.view.bounds.size.height);
    [self.current willMoveToParentViewController:nil];
    [self addChildViewController:new];
    new.view.frame = initialFrame;
    
    [self.current removeFromParentViewController];
    [new didMoveToParentViewController:self];
    self.current = new;
    if (completion) {
        completion();
    }
  }

- (UIStatusBarStyle) preferredStatusBarStyle {
    return UIStatusBarStyleLightContent;
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Authorization/RootParentVC.swift
//
//  RootParentVC.swift
//  sample-conference-videochat-swift
//
//  Created by Injoit on 1/27/20.
//  Copyright © 2020 quickBlox. All rights reserved.
//

import UIKit

class DialogsNavigationController: UINavigationController { }

class RootParentVC: UIViewController {
    var current: UIViewController
    var dialogID: String? {
        didSet {
            handlePush()
        }
    }
    
    init() {
        let storyboard = UIStoryboard(name: "Authorization", bundle: nil)
        current = storyboard.instantiateViewController(withIdentifier: "SplashScreenVC") as! SplashScreenVC
        super.init(nibName:  nil, bundle: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        addChild(current)
        current.view.frame = view.bounds
        view.addSubview(current.view)
        current.didMove(toParent: self)
    }
    
    func showLoginScreen() {
        let storyboard = UIStoryboard(name: "Authorization", bundle: nil)
        let authNavVC = storyboard.instantiateViewController(withIdentifier: "AuthNavVC") as! UINavigationController
        
        addChild(authNavVC)
        authNavVC.view.frame = view.bounds
        view.addSubview(authNavVC.view)
        authNavVC.didMove(toParent: self)
        
        current.willMove(toParent: nil)
        current.view.removeFromSuperview()
        current.removeFromParent()
        current = authNavVC
    }
    
    func goToDialogsScreen() {
        let storyboard = UIStoryboard(name: "Dialogs", bundle: nil)
        if let dialogsVC = storyboard.instantiateViewController(withIdentifier: "DialogsViewController") as? DialogsViewController {
            let dialogsScreen = DialogsNavigationController(rootViewController: dialogsVC)
            dialogsScreen.navigationBar.barTintColor = #colorLiteral(red: 0.2216441333, green: 0.4713830948, blue: 0.9869660735, alpha: 1)
            dialogsScreen.navigationBar.barStyle = .black
            dialogsScreen.navigationBar.shadowImage = UIImage()
            dialogsScreen.navigationBar.isTranslucent = false
            dialogsScreen.navigationBar.tintColor = .white
            dialogsScreen.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor: UIColor.white]
            
            addChild(dialogsScreen)
            dialogsScreen.view.frame = view.bounds
            view.addSubview(dialogsScreen.view)
            dialogsScreen.didMove(toParent: self)
            
            current.willMove(toParent: nil)
            current.view.removeFromSuperview()
            current.removeFromParent()
            current = dialogsScreen
        }
    }
    
    private func handlePush() {
        if let dialogsNavigationController = current as? DialogsNavigationController, let dialogID = self.dialogID {
            dialogsNavigationController.popToRootViewController(animated: false)
            (dialogsNavigationController.topViewController as? DialogsViewController)?.openChatWithDialogID(dialogID)
            self.dialogID = nil
        }
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }
}
